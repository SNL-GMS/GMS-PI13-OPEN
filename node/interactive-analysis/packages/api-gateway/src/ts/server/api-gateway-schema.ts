import { makeExecutableSchema } from 'apollo-server-express';
import merge from 'lodash/merge';
import path from 'path';
import { resolvers as cacheResolvers } from '../cache/resolvers';
import { schema as cacheSchema } from '../cache/schema';
import { resolvers as channelSegmentResolvers } from '../channel-segment/resolvers';
import { schema as channelSegmentSchema } from '../channel-segment/schema';
import {
  extendedResolvers as commonExtendedResolvers,
  resolvers as commonResolvers
} from '../common/resolvers';
import { extendedSchema as commonExtendedSchema, schema as commonSchema } from '../common/schema';
import { resolvers as configurationResolvers } from '../configuration/resolvers';
import { schema as configurationSchema } from '../configuration/schema';
import { resolvers as dataAcquisitionStatusResolvers } from '../data-acquisition/resolvers';
import { schema as dataAcquisitionStatusSchema } from '../data-acquisition/schema';
import { schema as eventSchema } from '../event/model-and-schema/schema';
import { resolvers as eventResolvers } from '../event/resolvers';
import { resolvers as fkResolvers } from '../fk/resolvers';
import { schema as fkSchema } from '../fk/schema';
import { resolvers as qcMaskResolvers } from '../qc-mask/resolvers';
import { schema as qcMaskSchema } from '../qc-mask/schema';
import { resolvers as signalDetectionResolvers } from '../signal-detection/resolvers';
import { schema as signalDetectionSchema } from '../signal-detection/schema';
import { resolvers as sohStatusResolvers } from '../soh/resolvers';
import { schema as sohStatusSchema } from '../soh/schema';
import { resolvers as processingStationResolvers } from '../station/processing-station/resolvers';
import { schema as processingStationSchema } from '../station/processing-station/schema';
import { resolvers as referenceStationResolvers } from '../station/reference-station/resolvers';
import { schema as referenceStationSchema } from '../station/reference-station/schema';
import { resolvers as systemMessageResolvers } from '../system-message/resolvers';
import { schema as systemMessageSchema } from '../system-message/schema';
import { resolvers as userProfileResolvers } from '../user-profile/resolvers';
import { schema as userProfileSchema } from '../user-profile/schema';
import { resolvers as waveformResolvers } from '../waveform/resolvers';
import { schema as waveformSchema } from '../waveform/schema';
import { resolvers as workflowResolvers } from '../workflow/resolvers';
import { schema as workflowSchema } from '../workflow/schema';

/**
 * Global GraphQL schema definition for the entire API gateway
 */

// using the gateway logger
import { gatewayLogger as logger } from '../log/gateway-logger';
const objectPath = path.relative(process.cwd(), __filename);

// GraphQL schema definitions
logger.info('Creating graphql schema...', { module: objectPath });

const typeDefs = [
  commonSchema,
  commonExtendedSchema,
  systemMessageSchema,
  workflowSchema,
  processingStationSchema,
  referenceStationSchema,
  waveformSchema,
  signalDetectionSchema,
  eventSchema,
  qcMaskSchema,
  userProfileSchema,
  channelSegmentSchema,
  fkSchema,
  dataAcquisitionStatusSchema,
  sohStatusSchema,
  cacheSchema,
  configurationSchema
];

// Merge GraphQL resolvers from the schema domains
const resolvers = merge(
  commonResolvers,
  commonExtendedResolvers,
  systemMessageResolvers,
  workflowResolvers,
  processingStationResolvers,
  referenceStationResolvers,
  waveformResolvers,
  signalDetectionResolvers,
  eventResolvers,
  userProfileResolvers,
  qcMaskResolvers,
  channelSegmentResolvers,
  fkResolvers,
  dataAcquisitionStatusResolvers,
  sohStatusResolvers,
  cacheResolvers,
  configurationResolvers
);

// Build the GraphQL schema
export const schema = makeExecutableSchema({
  typeDefs,
  resolvers
});
