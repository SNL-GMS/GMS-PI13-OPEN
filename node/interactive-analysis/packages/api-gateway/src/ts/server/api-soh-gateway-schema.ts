import { makeExecutableSchema } from 'apollo-server-express';
import merge from 'lodash/merge';
import path from 'path';
import { resolvers as commonResolvers } from '../common/resolvers';
import { schema as commonSchema } from '../common/schema';
import { resolvers as configurationResolvers } from '../configuration/resolvers';
import { schema as configurationSchema } from '../configuration/schema';
import { resolvers as sohStatusResolvers } from '../soh/resolvers';
import { schema as sohStatusSchema } from '../soh/schema';
import { resolvers as systemMessageResolvers } from '../system-message/resolvers';
import { schema as systemMessageSchema } from '../system-message/schema';
import { resolvers as userProfileResolvers } from '../user-profile/resolvers';
import { schema as userProfileSchema } from '../user-profile/schema';

/**
 * Global GraphQL schema definition for the entire API gateway
 */

// using the gateway logger
import { gatewayLogger as logger } from '../log/gateway-logger';
const objectPath = path.relative(process.cwd(), __filename);

// GraphQL schema definitions
logger.info('Creating graphql schema...', { module: objectPath });

const typeDefs = [
  commonSchema,
  userProfileSchema,
  systemMessageSchema,
  sohStatusSchema,
  configurationSchema
];

// Merge GraphQL resolvers from the schema domains
const resolvers = merge(
  commonResolvers,
  userProfileResolvers,
  systemMessageResolvers,
  sohStatusResolvers,
  configurationResolvers
);

// Build the GraphQL schema
export const schema = makeExecutableSchema({
  typeDefs,
  resolvers
});
