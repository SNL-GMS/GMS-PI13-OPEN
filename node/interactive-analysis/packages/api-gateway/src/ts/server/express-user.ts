/** defines an undefined user; used for the graphql playground */
export const UNDEFINED_USER = 'undefined';

/**
 * Represents a user
 */
export interface ExpressUser {
  userName: string;
}

/** Represents a map that maps a unique key to an ExpressUser */
export type ExpressUserMap = Map<string, ExpressUser>;
