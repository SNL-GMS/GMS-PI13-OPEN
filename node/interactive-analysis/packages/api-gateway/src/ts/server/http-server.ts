import config from 'config';
import { Express } from 'express';
import http from 'http';
import { gatewayLogger as logger } from '../log/gateway-logger';

/**
 * Creates the HTTP server
 * @param app the express server
 * @param ssl true if running in ssl; false otherwise
 */
export const createHttpServer = (app: Express): http.Server => {
  logger.info(`Creating the http server...`);

  // Load configuration settings
  const gqlConfig = config.get('server.graphql');

  // GraphQL Path
  const graphqlPath = gqlConfig.http.graphqlPath;
  logger.info(`graphqlPath ${graphqlPath}`);

  // GraphQL HTTP server port
  const httpPort = gqlConfig.http.port;
  logger.info(`httpPort ${httpPort}`);

  // GraphQL Websocket port
  const wsPort = gqlConfig.ws.port;
  logger.info(`wsPort ${wsPort}`);

  const server: http.Server =
    // Listen for GraphQL requests over HTTP
    http.createServer(app);

  server.listen(httpPort, function() {
    logger.info(`listening on port ${httpPort}`);
  });

  return server;
};
