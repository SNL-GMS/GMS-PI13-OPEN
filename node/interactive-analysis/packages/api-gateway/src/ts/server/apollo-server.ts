import { ApolloServer } from 'apollo-server-express';
import config from 'config';
import { Express } from 'express';
import { GraphQLError, GraphQLSchema } from 'graphql';
import msgpack from 'msgpack-lite';
import { CacheProcessor } from '../cache/cache-processor';
import { UserContext } from '../cache/model';
import { gatewayLogger as logger } from '../log/gateway-logger';
import { performanceLogger } from '../log/performance-logger';
import { ExpressUserMap, UNDEFINED_USER } from './express-user';

/**
 * Creates the Apollo Express Server
 * @param app the express server
 * @param schema the schema
 * @param userMap the user map
 */
export const createApolloExpressServer = (
  app: Express,
  schema: GraphQLSchema,
  userMap: ExpressUserMap
): ApolloServer => {
  logger.info(`Creating apollo server...`);

  // Load configuration settings
  const gqlConfig = config.get('server.graphql');

  // GraphQL Path
  const graphqlPath = gqlConfig.http.graphqlPath;
  logger.info(`graphqlPath ${graphqlPath}`);

  // GraphQL Websocket port
  const wsPort = gqlConfig.ws.port;
  logger.info(`wsPort ${wsPort}`);

  const apolloServer = new ApolloServer({
    schema,
    tracing: process.env.NODE_ENV !== 'production',
    cacheControl: true,
    // TODO: disable in production
    introspection: true, // process.env.NODE_ENV !== 'production',
    debug: process.env.NODE_ENV !== 'production',
    persistedQueries: false,
    context: ({ req }: { req: any }): UserContext => ({
      sessionId: req.sessionID,
      // TODO return string 'undefined' to avoid analyst.userName error in workflow
      // TODO - this case is hit when using firefox, more debugging is needed to get this fixed properly
      userName:
        userMap.has(req.sessionID) && userMap.get(req.sessionID).userName
          ? userMap.get(req.sessionID).userName
          : UNDEFINED_USER,
      userCache: userMap.has(req.sessionID)
        ? CacheProcessor.Instance().getCacheForUser(req.sessionID)
        : undefined,
      userRole: 'DEFAULT'
    }),
    playground: {
      endpoint: graphqlPath,
      subscriptionEndpoint: `ws://${gqlConfig.ws.host}:${wsPort}${gqlConfig.ws.path}`,
      settings: {
        'request.credentials': 'include'
      }
    },
    formatResponse: (response, options) => {
      // !IMPORTANT TODO - sometimes options.request is undefined
      // !there is no adequately explained reason that options.request isn't defined - look into this issue
      if (options && options.request && options.request.http && options.request.http.headers) {
        const headers = options.request.http.headers.get('Accept');
        if (headers !== null && headers.includes('application/msgpack')) {
          response.data = msgpack.encode(response.data);
          return response;
        }
      }
      // Example of data to extract data request out of
      // "data": {
      //     "clientLog": {
      //       "logLevel": "DATA",
      //       "message": "query defaultStations (in 3340 ms)",
      //       "time": "2019-03-28T14:36:55.285Z",
      //       "__typename": "ClientLog"
      //     }
      //   },
      // tslint:disable-next-line: no-for-in
      for (const property in response) {
        if (response.hasOwnProperty(property) && property === 'data') {
          const dataRequestName = Object.keys(response[property])[0];
          if (!dataRequestName.includes('clientLog')) {
            performanceLogger.performance(dataRequestName, 'returningFromServer');
          }
        }
      }
      return response;
    },
    formatError: (error: GraphQLError) => {
      logger.error(error.message);
      return error;
    }
  });

  apolloServer.applyMiddleware({ app });

  return apolloServer;
};
