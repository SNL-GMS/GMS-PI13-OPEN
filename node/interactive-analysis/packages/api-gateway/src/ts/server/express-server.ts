import bodyParser from 'body-parser';
import config from 'config';
import { Express } from 'express';
import session, { MemoryStore } from 'express-session';
import { v4 as uuid } from 'uuid';
import { gatewayLogger as logger } from '../log/gateway-logger';
import { ExpressUserMap } from './express-user';

/** Represents 24 hours */
export const TWENTY_FOUR_HOURS = 86400000; // 24 * 60 * 60 * 1000

/** The number of days to allow the session to stay active */
export const NUMBER_DAYS = 7;

/** The session lifetime */
export const SESSION_LIFETIME = TWENTY_FOUR_HOURS * NUMBER_DAYS;

/** The session name */
export const SESSION_NAME = 'InteractiveAnalysis';

/** True if running as ssl; false otherwise */
export const SSL = process.env.SSL === 'true';

/** Returns the protocol */
export const getProtocol = () => (SSL ? 'https://' : 'http://');

/**
 * Creates the Express Server
 * @param userMap the user map
 */
export const createExpressServer = (userMap: ExpressUserMap): Express => {
  logger.info(`Creating the express server...`);

  // tslint:disable-next-line:no-implicit-dependencies no-var-requires no-require-imports
  const express = require('express');
  const app: Express = express();

  app.set('trust proxy', true);

  const useInMemory: boolean = config.get('inMemorySession');

  app.use(
    session({
      name: SESSION_NAME,
      secret: 'upside down hedgehog',
      genid: req => uuid(), // use UUIDs for unique session IDs
      resave: true,
      store: useInMemory
        ? new MemoryStore()
        : // tslint:disable-next-line:no-implicit-dependencies no-var-requires no-require-imports
          new (require('connect-pg-simple')(session))({
            conString:
              'postgres://gms_session_application:gmsdb:postgres:gms_session_application:thawed-mortals-rebounds@postgresql-gms:5432/gms',
            tableName: 'session',
            schemaName: 'gms_session'
          }),

      saveUninitialized: true,
      proxy: true,
      cookie: {
        expires: new Date(Date.now() + SESSION_LIFETIME),
        maxAge: SESSION_LIFETIME,
        sameSite: true,
        secure: SSL
      }
    })
  );

  app.use(bodyParser.json({ limit: '50mb' }));
  app.use(bodyParser.urlencoded({ limit: '50mb', extended: true }));

  return app;
};
