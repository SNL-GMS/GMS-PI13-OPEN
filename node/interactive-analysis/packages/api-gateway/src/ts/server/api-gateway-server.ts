import { ConfigurationProcessor } from '../configuration/configuration-processor';
import { FkProcessor } from '../fk/fk-processor';
import { KafkaConsumer } from '../kafka/kafka-consumer';
import { KafkaProducer } from '../kafka/kafka-producer';
import { gatewayLogger as logger } from '../log/gateway-logger';
import { SohProcessor } from '../soh/soh-processor';
import { ProcessingStationProcessor } from '../station/processing-station/processing-station-processor';
import { ReferenceStationProcessor } from '../station/reference-station/reference-station-processor';
import { SystemMessageProcessor } from '../system-message/system-message-processor';
import * as waveformProcessor from '../waveform/waveform-processor';
import { WorkflowProcessor } from '../workflow/workflow-processor';
import { schema } from './api-gateway-schema';
import { createApolloExpressServer } from './apollo-server';
import { createExpressServer, getProtocol } from './express-server';
import { ExpressUser, ExpressUserMap } from './express-user';
import { createHttpServer } from './http-server';
import {
  configureRouteAlive,
  configureRouteAuthentication,
  configureRouteCheckApollo,
  configureRouteCheckInitialized,
  configureRouteCheckKafka,
  configureRouteCheckWebsocket,
  configureRouteHealthCheck,
  configureRouteReady
} from './routes';
import { createWebSocketServer } from './websocket-server';

logger.info('Starting API Gateway Server...');

const userMap: ExpressUserMap = new Map<string, ExpressUser>();
const app = createExpressServer(userMap);
createHttpServer(app);
configureRouteAlive(app);
configureRouteReady(app, getProtocol());
configureRouteHealthCheck(app, getProtocol());

const initializeProcessors = async () => {
  // Initialize the API Gateway Processors
  logger.info(`==> initialize processors, data, and initial configuration`);

  // ! The initialization order matters
  // Configuration Processor which makes a network call to get the Analyst UI Configuration.
  await ConfigurationProcessor.Instance().fetchConfiguration();
  await ReferenceStationProcessor.Instance().getDefaultStations();
  await ProcessingStationProcessor.Instance().fetchStationData();
  await WorkflowProcessor.Instance().fetchWorkflowData();

  configureRouteCheckInitialized(app);
};

const initializeKafka = async () => {
  // Initialize the KAFKA Consumers and Producers
  logger.info(`==> initialize KAFKA configurations`);

  // Initialize the KAFKA consumer
  await KafkaConsumer.Instance().start();

  // Initialize the SOH Kafka producer
  await KafkaProducer.Instance().start();

  // register callbacks for kafka
  SohProcessor.Instance().registerKafkaConsumerCallbacks();
  SystemMessageProcessor.Instance().registerKafkaConsumerCallbacks();

  configureRouteCheckKafka(app);
};

const initializeApollo = () => {
  logger.info(`==> initialize Apollo Express server`);
  createApolloExpressServer(app, schema, userMap);
  configureRouteCheckApollo(app, getProtocol());
};

const initializeWebsocketServer = () => {
  logger.info(`==> initialize websocket server`);
  createWebSocketServer(schema, userMap);
  configureRouteCheckWebsocket(app);
};

initializeProcessors()
  .then(initializeKafka)
  .then(initializeApollo)
  .then(initializeWebsocketServer)
  .then(() => {
    // Calling because in mock the loading of the Fk Channel Segments
    // take long enough to timeout UI call to compute Fk
    FkProcessor.Instance(); // Testing fk loading

    logger.info('register /waveforms/raw');
    app.get('/waveforms/raw', waveformProcessor.waveformRawSegmentRequestHandler);

    logger.info('register /waveforms/filter');
    app.get('/waveforms/filtered', waveformProcessor.waveformFilteredSegmentRequestHandler);

    configureRouteAuthentication(app, userMap);
  })
  .catch(e => logger.error(e));
