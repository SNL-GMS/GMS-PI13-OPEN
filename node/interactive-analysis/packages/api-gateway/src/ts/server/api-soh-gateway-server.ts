// tslint:disable: no-magic-numbers

import { ConfigurationProcessor } from '../configuration/configuration-processor';
import { KafkaConsumer } from '../kafka/kafka-consumer';
import { KafkaProducer } from '../kafka/kafka-producer';
import { gatewayLogger, gatewayLogger as logger } from '../log/gateway-logger';
import { SohProcessor } from '../soh/soh-processor';
import { ProcessingStationProcessor } from '../station/processing-station/processing-station-processor';
import { SystemMessageProcessor } from '../system-message/system-message-processor';
import { schema } from './api-soh-gateway-schema';
import { createApolloExpressServer } from './apollo-server';
import { createExpressServer, getProtocol } from './express-server';
import { ExpressUser, ExpressUserMap } from './express-user';
import { createHttpServer } from './http-server';
import {
  configureRouteAlive,
  configureRouteAuthentication,
  configureRouteCheckApollo,
  configureRouteCheckInitialized,
  configureRouteCheckKafka,
  configureRouteCheckWebsocket,
  configureRouteHealthCheck,
  configureRouteReady
} from './routes';
import { createWebSocketServer } from './websocket-server';

logger.info('Starting API SOH Gateway Server...');

const userMap: ExpressUserMap = new Map<string, ExpressUser>();
const app = createExpressServer(userMap);
createHttpServer(app);
configureRouteAlive(app);
configureRouteReady(app, getProtocol());
configureRouteHealthCheck(app, getProtocol());

const initializeProcessors = async () => {
  try {
    // Initialize the API Gateway Processors
    logger.info(`==> initialize processors, data, and initial configuration`);

    // ! The initialization order matters
    // Configuration Processor which makes a network call to get the Analyst UI Configuration.
    await ConfigurationProcessor.Instance().fetchConfiguration();
    await ProcessingStationProcessor.Instance().fetchStationData();
    configureRouteCheckInitialized(app);
  } catch (error) {
    gatewayLogger.error(`Failed to initialize processors: ${error}`);
  }
};

const initializeKafka = async () => {
  try {
    // Initialize the KAFKA Consumers and Producers
    logger.info(`==> initialize KAFKA configurations`);

    // Initialize the system message Kafka consumer
    await KafkaConsumer.Instance().start();

    // Initialize the SOH Kafka producer
    await KafkaProducer.Instance().start();

    // register callbacks for kafka
    SohProcessor.Instance().registerKafkaConsumerCallbacks();
    SystemMessageProcessor.Instance().registerKafkaConsumerCallbacks();

    configureRouteCheckKafka(app);
  } catch (error) {
    gatewayLogger.error(`Failed to initialize KAFKA: ${error}`);
  }
};

const initializeApollo = () => {
  try {
    logger.info(`==> initialize Apollo Express server`);
    createApolloExpressServer(app, schema, userMap);
    configureRouteCheckApollo(app, getProtocol());
  } catch (error) {
    gatewayLogger.error(`Failed to initialize Apollo: ${error}`);
  }
};

const initializeWebsocketServer = () => {
  try {
    logger.info(`==> initialize websocket server`);
    createWebSocketServer(schema, userMap);
    configureRouteCheckWebsocket(app);
  } catch (error) {
    gatewayLogger.error(`Failed to initialize Websocket Server: ${error}`);
  }
};

initializeProcessors()
  .then(initializeKafka)
  .then(initializeApollo)
  .then(initializeWebsocketServer)
  .then(() => {
    configureRouteAuthentication(app, userMap);
  })
  .catch(e => logger.error(e));
