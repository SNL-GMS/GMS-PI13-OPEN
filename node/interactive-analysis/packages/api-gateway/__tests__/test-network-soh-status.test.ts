import { graphql } from 'graphql';
import { schema } from '../src/ts/server/api-gateway-schema';

// ---- Query test cases ----

// Test case - basic content query for network SOH status
// Query for network SOH status by time range
it('Network SOH Status query should match snapshot', async () => {
  // language=GraphQL
  const sohStatusQuery = `
    query networkSohStatus($startTime: String, $endTime: String) {
      networkSohStatus(startTime: $startTime, endTime: $endTime) {
        id
        networkId
        networkName
        startTime
        endTime
        sohStatusSummary
        stationSohStatus [{
          stationId
          stationName
          sohStatusSummary
          sohStatus {
            stationAcquisitionSohStatus {
              completeness
              completenessSummary
              lag
              lagSummary
            }
            environmentSohStatus {
              countBySohType
              summaryBySohType
            }
          }
          stationSohIssue {
            requiresAcknowledgement
            acknowledgedAt
          }
          channelSohStatus {
            channelId
            channelName
            sohStatus {
              stationAcquisitionSohStatus {
                completeness
                completenessSummary
                lag
                lagSummary
              }
              environmentSohStatus {
                countBySohType
                summaryBySohType
              }
            }
          }
        }]
      }
    }
  `;

  const rootValue = {};
  // Execution the GraphQL query with a small delay to allow the API gateway to settle async
  // HTTP requests fetching data from the mock backend
  const result = await graphql(schema, sohStatusQuery, rootValue);

  const { data } = result;

  // Compare response to snapshot
  expect(data).toMatchSnapshot();
});
