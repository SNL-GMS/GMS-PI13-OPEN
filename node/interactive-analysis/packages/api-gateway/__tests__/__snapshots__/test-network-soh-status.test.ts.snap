// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Network SOH Status query should match snapshot 1`] = `undefined`;
